live-debugger-ffi/src/data.rs

```diff
@@ -0,0 +1,297 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache
+// License Version 2.0. This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+
+use datadog_live_debugger::debugger_defs::{ProbeMetadata, ProbeMetadataLocation};
+use datadog_live_debugger::{
+    CaptureConfiguration, DslString, EvaluateAt, InBodyLocation, MetricKind, ProbeCondition,
+    ProbeValue, SpanProbeTarget,
+};
+use ddcommon_ffi::slice::AsBytes;
+use ddcommon_ffi::{CharSlice, Option};
+
+#[repr(C)]
+pub struct CharSliceVec<'a> {
+    pub strings: *const CharSlice<'a>,
+    pub string_count: usize,
+}
+
+impl<'a> Drop for CharSliceVec<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            Vec::from_raw_parts(
+                self.strings as *mut CharSlice,
+                self.string_count,
+                self.string_count,
+            )
+        };
+    }
+}
+
+impl<'a> From<&'a Vec<String>> for CharSliceVec<'a> {
+    fn from(from: &'a Vec<String>) -> Self {
+        let char_slices: Vec<CharSlice<'a>> = from.iter().map(|s| s.as_str().into()).collect();
+        let new = CharSliceVec {
+            strings: char_slices.as_ptr(),
+            string_count: char_slices.len(),
+        };
+        std::mem::forget(char_slices);
+        new
+    }
+}
+
+#[repr(C)]
+pub struct MetricProbe<'a> {
+    pub kind: MetricKind,
+    pub name: CharSlice<'a>,
+    pub value: &'a ProbeValue,
+}
+
+impl<'a> From<&'a datadog_live_debugger::MetricProbe> for MetricProbe<'a> {
+    fn from(from: &'a datadog_live_debugger::MetricProbe) -> Self {
+        MetricProbe {
+            kind: from.kind,
+            name: from.name.as_str().into(),
+            value: &from.value,
+        }
+    }
+}
+
+#[repr(C)]
+pub struct LogProbe<'a> {
+    pub segments: &'a DslString,
+    pub when: &'a ProbeCondition,
+    pub capture: &'a CaptureConfiguration,
+    pub capture_snapshot: bool,
+    pub sampling_snapshots_per_second: u32,
+}
+
+impl<'a> From<&'a datadog_live_debugger::LogProbe> for LogProbe<'a> {
+    fn from(from: &'a datadog_live_debugger::LogProbe) -> Self {
+        LogProbe {
+            segments: &from.segments,
+            when: &from.when,
+            capture: &from.capture,
```

Comment:
What is the reason for these fields being pointers?

Reply:
Avoiding unnecessary copies - this field is directly from an exportable rust representation and thus doesn't need a copy.